metatheory/TestNEAT.hs

```diff
@@ -0,0 +1,105 @@
+module Main where
+
+import           Control.Monad.Except
+import           Data.Coolean
+import           Data.Either
+import           Language.PlutusCore
+import           Language.PlutusCore.Generators.NEAT.PropTest
+import           Language.PlutusCore.Normalize
+import           Test.Tasty
+import           Test.Tasty.HUnit
+
+import           MAlonzo.Code.Main                            (checkKindAgda, inferKindAgda, normalizeTypeAgda)
+import           MAlonzo.Code.Scoped                          (deBruijnifyK, unDeBruijnifyK)
+
+import           Language.PlutusCore.DeBruijn
+import           Raw
+
+
+main :: IO ()
+main = defaultMain allTests
+
+allTests :: TestTree
+allTests = testGroup "all tests"
+  [ testCaseCount "soundness" $
+      testTyProp depth kind prop_checkKindSound
+  , testCaseCount "normalization" $
+      testTyProp depth kind prop_normalizePreservesKind
+  , testCaseCount "normalizationSound" $
+      testTyProp depth kind prop_normalizeTypeSound
+  , testCaseCount "normalizationAgree" $
+      testTyProp depth kind prop_normalizeTypeSame
+  , testCaseCount "kindInferAgree" $
+      testTyProp depth kind prop_kindInferSame
+  ]
+
+testCaseCount :: String -> IO Integer -> TestTree
+testCaseCount name act = testCaseInfo name $
+  liftM (\i -> show i ++ " types generated") act
+
+
+-- NEAT settings
+
+depth :: Int
+depth = 10
+
+kind :: Kind ()
+kind = Type ()
+
+-- |Check if the type/kind checker or generation threw any errors.
+isSafe :: ExceptT (ErrorP a) Quote a -> Cool
+isSafe = toCool . isRight . runQuote . runExceptT
+
+prop_checkKindSound :: TyProp
+prop_checkKindSound k _ tyQ = isSafe $ do
+  ty <- withExceptT GenErrorP tyQ
+  tyDB <- withExceptT FVErrorP $ deBruijnTy ty
+  withExceptT TypeErrorP $ case checkKindAgda (AlexPn 0 0 0 <$ tyDB) (deBruijnifyK (convK k)) of
```

Comment:
Here we run `withExceptT` sequentially and then lift everything into the concurrent `Cool` via `isSafe`. Would it make more sense to generalize `isSafe` to work over any error, call `isSafe` for each of the checks (without calling `withExceptT`) and then fold the results with `Data.Coolean.&&&` in order to parallelize the checks?

On the other hand given that we expect the tests to succeed, there's not much point in optimizing the non-happy path as that would likely make the happy path less efficient.

Maybe worth mentioning this in a Note somewhere.